consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/validation/MessageValidator.java

```diff
@@ -0,0 +1,194 @@
+    this.expectedProposer = expectedProposer;
+    this.roundIdentifier = roundIdentifier;
+    this.headerValidator = headerValidator;
+    this.protocolContext = protocolContext;
+    this.parentHeader = parentHeader;
+  }
+
+  public boolean addPreprepareMessage(
+      final IbftSignedMessageData<IbftUnsignedPrePrepareMessageData> msg) {
+
+    if (preprepareMessage.isPresent()) {
+      return handleSubsequentPreprepareMessage(msg);
+    }
+
+    if (!msg.getUnsignedMessageData().getRoundIdentifier().equals(roundIdentifier)) {
+      LOG.info("Invalid Preprepare message, does not match current round.");
+      return false;
+    }
+
+    if (preprepareMessage.isPresent()) {
+      LOG.info("Invalid Preprepare message, one already exists for the current round.");
+      return false;
+    }
+
+    if (!msg.getSender().equals(expectedProposer)) {
+      LOG.info(
+          "Invalid Preprepare message, was not created by the proposer expected for the "
+              + "associated round.");
+      return false;
+    }
+
+    final Block proposedBlock = msg.getUnsignedMessageData().getBlock();
+    if (!headerValidator.validateHeader(
+        proposedBlock.getHeader(), parentHeader, protocolContext, FULL)) {
+      LOG.info("Invalid Prepare message, block did not pass header validation.");
+      return false;
+    }
+
+    preprepareMessage = Optional.of(msg);
+    return true;
+  }
+
+  private boolean handleSubsequentPreprepareMessage(
+      final IbftSignedMessageData<IbftUnsignedPrePrepareMessageData> msg) {
+    IbftSignedMessageData<IbftUnsignedPrePrepareMessageData> existingMsg = preprepareMessage.get();
+
+    if (!existingMsg.getSender().equals(msg.getSender())) {
+      LOG.debug("Received subsequent invalid Preprepare message; sender differs from original.");
+      return false;
+    }
+
+    final IbftUnsignedPrePrepareMessageData existingData = existingMsg.getUnsignedMessageData();
+    final IbftUnsignedPrePrepareMessageData newData = msg.getUnsignedMessageData();
+
+    if (!existingData.equals(newData)) {
+      LOG.debug("Received subsequent invalid Preprepare message; content differs from original.");
+      return false;
+    }
+
+    return true;
+  }
+
+  public boolean validatePrepareMessage(
+      final IbftSignedMessageData<IbftUnsignedPrepareMessageData> msg) {
+    final String msgType = "Prepare";
+
+    if (!isMessageInRoundFromValidatorAndMatchesPreprepareDigest(msg, msgType)) {
+      return false;
+    }
+
+    if (msg.getSender().equals(expectedProposer)) {
+      LOG.info("Illegal Prepare message; was sent by the round's proposer.");
+      return false;
+    }
+
+    return validateDigestMatchesPreprepareBlock(msg.getUnsignedMessageData().getDigest(), msgType);
+  }
+
+  public boolean validateCommmitMessage(
+      final IbftSignedMessageData<IbftUnsignedCommitMessageData> msg) {
+    final String msgType = "Commit";
+
+    if (!isMessageInRoundFromValidatorAndMatchesPreprepareDigest(msg, msgType)) {
```

Comment:
I think commit is only meant to validate that round number > current round number and not check the sequence number. Not sure if it matters or not.

Reply:
I'm getting the impression that the commit message's consensusroundID needs to match the associated Preprepare message - so .... take it up with Roberto.

Reply:
all good as we are discarding old commit messages